### PR TITLE
Allow worker's group and development passwords.

### PR DESCRIPTION
**Pull request by vvereschaka**
This is an improvement for 'create_worker()' function, which allows
to use a single group password for a set of the workers instead of
a worker specific password.

Also it provides a default development password for the workers with
no password specified in case the buildbot instance is configured as
a development instance.

### CODE REVIEW

File: buildbot/osuosl/master/config/workers.py [hunk 1]
```diff
@@ -1,12 +1,59 @@
...
+    is_prod = config.options.getboolean('Master Options', 'is_production', fallback=False)
+    is_dev = config.options.getboolean('Master Options', 'is_dev', fallback=False)
+    password_fallback = "no_pass" if is_dev else None
+
+    if group:
+        password = config.options.get(password_section, group,
+                                      fallback=config.options.get(password_section, name, fallback=password_fallback))
+    else:
+        password = config.options.get(password_section, name, fallback=password_fallback)
+
```
Comment (slydiman) [inline buildbot/osuosl/master/config/workers.py:49]: Consider
```
password = config.options.get(password_section, group or "no_group", 
     fallback=config.options.get(password_section, name, fallback=password_fallback))
```

**Review by slydiman — APPROVED** [review]
LGTM

File: buildbot/osuosl/master/config/workers.py (hunk 1 above)
Comment (vvereschaka) [reply buildbot/osuosl/master/config/workers.py:49]: I think we should not support the "no_group" entry for the password section. If there is no group name then use the worker name. I.e. we don't need a "default" group name there.

Comment (slydiman) [reply buildbot/osuosl/master/config/workers.py:49]: `no_group` is a dummy name (must be missing in the config) to use fallback if group is None. It is a way to optimize code and remove if/else with the same `config.options.get(password_section, name, fallback=password_fallback)`. Up to you, of course.

Comment (vvereschaka) [reply buildbot/osuosl/master/config/workers.py:49]: got it, looks like a good idea. I'll update the changes.

Comment (vvereschaka) [reply buildbot/osuosl/master/config/workers.py:49]: updated

**Review by gkistanova — COMMENTED** [review]
The patch itself looks good.

I have a concern related to this patch, though. Currently each worker gets added individually, gets staged and then promoted to production only if approved (I check the worker description, owner information, and build history before I approve the move). A worker password is a way to allow or not allow a worker into production.

With group password we are loosing a way to validate and approve individual workers, which likely would lead to missing important information in contact info and/or worker description.

Any idea how to address this and have a convenience of having group worker access?

**Comment by vvereschaka** [comment]
@gkistanova,
>Any idea how to address this and have a convenience of having group worker access?

Hm, I though you approve the workers by approving the changes within `llvm-zorg/buildbot/osuosl/master/config/workers.py` file. Is there a case when someone adds a new worker there, you approve it, but you don't want this worker to get active?

**Comment by gkistanova** [comment]
> Hm, I though you approve the workers by approving the changes within llvm-zorg/buildbot/osuosl/master/config/workers.py file. Is there a case when someone adds a new worker there, you approve it, but you don't want this worker to get active?

The actual procedure is:

1. A patch for a new builder and worker gets introduced for review. At the review I check if the code makes sense, free of obvious problems, adds a valuable build configuration, and such. If everything Ok, I approve.
2. Once this patch is in, I add worker/password on the staging. So, owner could connect his worker to staging and polish it, resolve issues, make it green. I do not look closely at the staged workers up to the point when owner claims it is ready for production.  At this step the workers in question could be used on staging but not in production. If owner would point such a worker to production it will not connect.
3. After the owner said it is ready, I check information configured on the worker side, such as owner contact info and worker description, as well as the build history to see if the builder is stable, reliable and reasonably fast. If everything is Ok, I add worker credentials to production and let owner know that he could move the worker there. If there are problems, owner would have to fix them before it is allowed in production.

I prefer having a control and check new workers before they would go to production. Often owners left buildbot default descriptions, put some questionable things there, or ignore issues rushing to production. 

Hope it makes sense.

**Comment by vvereschaka** [comment]
Thank you @gkistanova .
Now I understood the workflow. It makes sense. I'll think a bit more about it

**Pull request closed without merging**